### DISCUSSION TITLE
Swagger UI not showing optional file upload button

### DISCUSSION THREAD

**Discussion by jvdavim** [fastapi/fastapi#10755] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.post("/example")
def read_item(upload_file: Annotated[UploadFile | None, File(description="Some description")] = None):
    return
```

### Description

~~For some reason, when using Python >= 3.10 optional type hints ( `SomeType | None` ) in endpoints methods parameters, the upload button is not rendered in the Swagger UI.~~
Edit: For some reason, when making an `UploadFile` parameter optional, the upload button is not rendered in the Swagger UI. The problem is just the same using the old `typing.Optional` notation or `UploadFile | None`.

Update: It's happening only with pydantic v2. Downgrading to pydantic v1 seems to fix the bug.

An example of working swagger UI (but wrong typing) upload button:
![2023-12-04_17-39](https://github.com/tiangolo/fastapi/assets/16657663/6562898e-cb10-49f3-a23b-d74890c089c7)
![2023-12-04_17-38](https://github.com/tiangolo/fastapi/assets/16657663/853911a8-6207-49d6-91b4-929ed05490dc)

An example of broken swagger UI (but right typing) upload button:
![2023-12-04_17-37](https://github.com/tiangolo/fastapi/assets/16657663/8b587830-b146-48eb-b9b5-1441df8b9e48)
![2023-12-04_17-36](https://github.com/tiangolo/fastapi/assets/16657663/c909aff3-cf90-4a16-950e-b46ffb1ed781)

Inspecting the generated openapi.json I realized that there are some differences on using the optional (`SomeType | None`) type hint, as seem on the screenshot below:
![2023-12-04_17-49](https://github.com/tiangolo/fastapi/assets/16657663/dbfaa764-6f0d-4f32-a7ae-34e70022353f)

I'm not sure if this change is not expected, or if it's a Swagger UI related problem. I tried to find it in Swagger UI repository, but I was not successful. I'm not so experienced with openapi docs, so maybe I'm missing something. I appreciate your help.

### Operating System

Linux, macOS

### FastAPI Version

0.104.1

### Pydantic Version

2.5.2

### Python Version

3.11.6

**Comment by kiraware** [comment]
```python
from typing import Optional

from fastapi import FastAPI
from pydantic import BaseModel


app = FastAPI()


class Blog(BaseModel):
    title: str
    body: str
    published: Optional[bool] # published: bool | None, should just do the same thing


@app.post("/test")
async def create_blog(request: Blog):
    return request

```

This code too. Optional is fail and the published field is required in Swagger UI docs.

![image](https://github.com/tiangolo/fastapi/assets/117554978/e669b2d6-2343-4e05-8fd7-5f3227ccd27f)

Or redocs

![image](https://github.com/tiangolo/fastapi/assets/117554978/f1098b39-f1f7-4502-8ee1-9505571861e9)

Am i miss something?

**Reply by sr2ds** [reply]
Looks that when the field is Optional for Upload type, the button is not showed on UI

**Reply by jvdavim** [reply]
You are both right. It seems not work with both `typing.Optional` and `UploadFile | None` notations. So it's not right having Python 3.10 or the new type hints notation as causes of the problem. I'm gonna edit the original post.

**Comment by sr2ds** [comment]
@jvdavim , did you try with no `Annotated`?:

`def read_item(upload_file: UploadFile | None) `

**Reply by jvdavim** [reply]
I didn't. But I'm testing right now...
It seems the problem persists.
Using `def read_item(upload_file: UploadFile | None)` I get Almost the same result, but without the "send empty" checkbox:
![image](https://github.com/tiangolo/fastapi/assets/16657663/dfa480a1-d02b-4cb8-ac21-e38a85806f36)

Using `def read_item(upload_file: UploadFile | None = None)` I get the checkbox back to the SwaggerUI, but still missing button:
![image](https://github.com/tiangolo/fastapi/assets/16657663/cedc16b7-cc2f-4edb-b16c-f2dfa0920cb4)

**Comment by kiraware** [comment]
After some try i can make it optional. It is happen if you assign None to field in Pydantic model

```Python
...

class Blog(BaseModel):
    title: str
    body: str
    published: Optional[bool] = None # Or do with this, published: bool | None = None

...
```

The key is when you assign None to Pydantic model to make it optional in Swagger or redocs

```Python

class Foo(BaseModel):
    a: Optional[str] # Required
    b: str | None # Required
    c: Optional[str] | None # Required
    d: Optional[str] = None # Optional
    e: str | None = None # Optional
    f: str = None # Optional
```

**Reply by jvdavim** [reply]
Yes. Assigning `None` to the argument will definitely make it optional. Even so I have 2 considerations:
1. In your example the attribute `f` of the BaseModel is typed as `str`, but it's being set as `None`. The linter will show an error, because it's being set a type to the attribute which were not supposed to. So, to make it right, it should be explicitly marked as `str | None`.
2. I don't know yet if this behavior is expected, but IMHO making an `UploadFile` argument optional, shouldn't make the upload button disappear from the SwaggerUI.

**Comment by iudeen** [comment]
yes, I agree this is an issue!

**Comment by jvdavim** [comment]
I have found that it's happening only with pydantic v2. But it seems to be a problema with SwaggerUI, not FastAPI. I'm gonna try to find some issue on their repository or open a new one.

For some reason, the upload file button is not being rendered beacause SwaggerUI is not processing the `AnyOf` from OpenAPI 3.1 specification. Here is the swagger online editor: https://editor-next.swagger.io/. The OpenAPI 3.1.0 json produces the same result on the online editor.

**Comment by carlodri** [comment]
I have just stumbled on the same issue using fastapi 0.109.2 and pydantic 2.6.1

**Comment by niking00** [comment]
Use 
```python
@app.post("/example")
def read_item(upload_file: Annotated[UploadFile, File(description="Some description")] = None):
    return
```

Works for me

**Reply by jvdavim** [reply]
Yes. This works, but as I described, it's using wrong type hint. How could an `UploadFile` be `None` by default? It should be `UploadFile | None`.

**Comment by bpkrishnan** [comment]
Any update?

**Comment by xiaoyuan-sun** [comment]
Met same problem, any update?

**Comment by victor-rayan** [comment]
same problem here. The upload button only appears in swagger when UploadFile is required.

If i try 
   photo: UploadFile = File(None)

receive -> pydantic_core._pydantic_core.ValidationError: 1 validation error for PropertyCreateRequest
photo
  Value error, Expected UploadFile, received: <class 'NoneType'> [type=value_error, input_value=None, input_type=NoneType]

**Comment by YuriiMotov** [comment, marked as answer]
You can make it working using `SkipJsonSchema`.

The following solution works:

```python
from typing import Annotated

from fastapi import FastAPI, File, UploadFile
from pydantic.json_schema import SkipJsonSchema

app = FastAPI()

@app.post("/example")
def read_item(
    upload_file: Annotated[
        UploadFile | SkipJsonSchema[None], File(description="Some description")
    ] = None,
):
    print(upload_file)

```

There is [another issue](https://github.com/fastapi/fastapi/discussions/12150) here: if you try sending request using Swagger with "Send empty value" option, it will lead to validation error.
Browser sends empty value and FastAPI treats it as a string.
As a workaround you can add `BeforeValidator` that converts empty string to `None`:

```python
from typing import Annotated, Any

from fastapi import FastAPI, File, UploadFile
from pydantic import BeforeValidator
from pydantic.json_schema import SkipJsonSchema

app = FastAPI()

def empty_str_to_none(v: Any) -> Any:
    if v == "":
        return None
    return v

@app.post("/example")
def read_item(
    upload_file: Annotated[
        UploadFile | SkipJsonSchema[None],
        File(description="Some description"),
        BeforeValidator(empty_str_to_none),
    ] = None,
):
    print(upload_file)
```

**Comment by play123** [comment]
I’m experiencing the same issue. I tried your solution and it works correctly for a single file, but it does not work when using a list of files.

I also tested with different OpenAPI schemas, including Swagger UI, ReDoc, and Scalar, and the issue persists. It appears that this is not a FastAPI problem, but rather a Swagger UI rendering issue. When receiving file inputs, it displays a text field instead of a file selector.

I was able to resolve it by modifying the OpenAPI schema to use a binary format, but that feels like a Swagger-side workaround. Is there any alternative solution that does not require modifying the generated OpenAPI schema?

here is my code

from fastapi.openapi.utils import get_openapi


from fastapi import FastAPI, UploadFile, File
from typing import List

app = FastAPI()


@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile = File(...)):
    return {"filename": file.filename}


@app.post("/uploadfiles/")
async def create_upload_files(file: List[UploadFile] = File(...)):
    return {"filenames": [f.filename for f in file]}


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema

    schema = get_openapi(
        title=app.title,
        version=app.version,
        routes=app.routes,
    )

    # Force OAS 3.0.3 for best Swagger UI file upload support
    schema["openapi"] = "3.0.3"

    # Patch: convert OAS3.1 contentMediaType -> OAS3.0 format=binary
    components = schema.get("components", {}).get("schemas", {})
    for _, s in components.items():
        props = s.get("properties", {})
        for _, p in props.items():
            # Handle array of files: {"type":"array","items":{...}}
            if p.get("type") == "array" and isinstance(p.get("items"), dict):
                items = p["items"]
                if "contentMediaType" in items:
                    items.pop("contentMediaType", None)
                    items["format"] = "binary"
                    items["type"] = "string"
            # Handle single file: {"type":"string", ...}
            if p.get("type") == "string" and "contentMediaType" in p:
                p.pop("contentMediaType", None)
                p["format"] = "binary"

    app.openapi_schema = schema
    return app.openapi_schema


app.openapi = custom_openapi

**Reply by YuriiMotov** [reply]
You are probably experiencing the issue reported here: https://github.com/fastapi/fastapi/discussions/14975
We are looking at it
